src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+    instruction_ref new_split   = p_program->insert_instruction(
+        split_ins, op::split{axis, slice_dims, {enum_ndx, enum_ndx}}, input);
+    operation& op_edit = split_ins->get_operator_edit();
+
+    if(first == enum_ndx)
+        (any_cast<op::split>(op_edit)).slice_selector.first = enum_ndx + 1;
+    else
+        (any_cast<op::split>(op_edit)).slice_selector.second = enum_ndx - 1;
+
+    std::vector<shape> shapes;
+    shapes.push_back(input->get_shape());
+    shape new_shape = (any_cast<op::split>(op_edit)).compute_shape(shapes);
+    split_ins->set_shape(new_shape);
+    return new_split;
+}
+
+bool horizontal_fusion_impl::has_unique_output(const std::vector<instruction_ref>& instrs)
+{
+    std::unordered_map<instruction_ref, bool> seen;
+    for(auto&& ins : instrs)
+    {
+        for(auto&& output : ins->outputs())
+        {
+            if(seen.find(output) != seen.end())
+                return false;
+            seen[output] = true;
+        }
+    }
+    return true;
+}
+
+std::vector<unsigned> horizontal_fusion_impl::find_cluster(unsigned id)
+{
+    std::vector<unsigned> cluster;
+    cluster.push_back(id);
+    unsigned cur = id;
+    int size     = hash_instrs[id].size();
+    // Find a sub-tree of the hash tree to be fused together.
+    // Every node in the sub-tree contain the same amount of instructions.
+    while((hash_outputs.find(cur) != hash_outputs.end()) && (hash_outputs[cur].size() == 1))
+    {
+        unsigned output = (*(hash_outputs[cur].begin()))->id;
+        // Currently instruction can not have more than one same outputs.
+        // Therefore skip outputs that are not unique.
+        if((hash_instrs.find(output) != hash_instrs.end()) &&
+           (hash_instrs[output].size() == size) && has_unique_output(get_instrs(output)))
+        {
+            cluster.push_back(output);
+            cur = output;
+        }
+        else
+            break;
+    }
+    return cluster;
+}
+
+void horizontal_fusion_impl::transform_layers(
+    std::vector<std::vector<instruction_ref>>& all_inputs,
+    std::unordered_map<instruction_ref, instruction_ref>& root,
+    int axis,
+    std::vector<instruction_ref>& base_instrs)
+{
+    std::vector<instruction_ref> input0 = all_inputs.at(0);
+    // concat inputs.
+    for(unsigned long ndx = 0; ndx < input0.size(); ndx++)
+    {
+        std::vector<instruction_ref> instrs;
+        std::transform(all_inputs.begin(),
+                       all_inputs.end(),
+                       std::back_inserter(instrs),
+                       [&](auto&& d) -> instruction_ref { return d.at(ndx); });
+        concat(instrs, root, axis);
+    }
+
+    remove_redundant_roots(base_instrs);
+
+    // remove redundant inputs.
+    int ndx = 0;
+    for(auto&& input : all_inputs)
+    {
+        for(auto&& ins : input)
+        {
+            bool is_literal = (ins->name() == "@literal");
+            if((ndx == 0) && !is_literal)
```

Comment:
This condition should be moved out of the loop:

```cpp
        bool is_literal = (ins->name() == "@literal");
        auto start = is_literal ? input.begin() : input.begin()+1;
        for(auto&& ins : range(start, input.end()))
        {
            p_program->remove_instruction(ins);
        }
```

Reply:
not the same sematics.

Reply:
I see, its not that clear. This could be written as:

```cpp
        auto start = std::find_if(input.begin(), input.end(), [](auto ins) {
            return ins->name() == "@literal";
        });
        for(auto&& ins : range(start, input.end()))
        {
            p_program->remove_instruction(ins);
        }
```

Can you explain why you skip elements that are not literals at the beginning?

Reply:
keep the 1st chain, remove others.